backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchKnnPredicate.java

```diff
@@ -226,21 +237,37 @@ protected JsonObject doToJsonQuery(PredicateRequestContext context, JsonObject o
 			if ( filter != null ) {
 				FILTER_ACCESSOR.set( innerObject, filter );
 			}
-			K_ACCESSOR.set( innerObject, k );
+			if ( similarity != null ) {
+				MAX_DISTANCE.set( innerObject, similarity );
+			}
+			if ( score != null ) {
+				MIN_SCORE.set( innerObject, score );
+			}
+			if ( similarity == null && score == null ) {
+				// [knn] requires exactly one of k, distance or score to be set
+				K_ACCESSOR.set( innerObject, k );
+			}
```

Comment:
This may be a change of "interest" 😃 
as per that comment, OpenSearch only allows setting one of the options. Having `k` sent along with score/distance leads to an exception returned by the cluster. 
Hence my thinking here - if we have score/distance filter we ignore the `k` value and mention this behavior in the ref docs.